### ISSUE TITLE
Windows installer bug

### ISSUE THREAD

**Issue by Neustradamus** [sqlitebrowser/sqlitebrowser#1157]
It is possible to add a removal of previous version before start install of the new version?

**Comment by justinclift** [comment]
It should already be offering to remove the previous version when you install a new one on windows.

Is it not doing that for you?

**Comment by Neustradamus** [comment]
no, I had an old version, I forgot the version :/

**Comment by justinclift** [comment]
k.  Our installer definitely does check if there's an existing version, and offers to remove it if found.

However, it's possible the older version was **so** old that the installer didn't realise it.

Um... not really sure how to address this.  Thinking it's probably something we can skip looking into, unless people report it happening for modern versions (eg 3.8.0 onwards maybe?).

Thoughts? :smile:

**Comment by Neustradamus** [comment]
it was 3.8.1 maybe

**Comment by justinclift** [comment]
Ahhh, ok.  In that case we should probably look into it. :smile:

**Comment by chrisjlocke** [comment]
I've just installed VMWare so can set up a Win 8 environment (not sure which version @Neustradamus was using, but 8 seems the 'best of the bunch' to test with) to install 3.8.1 and then 3.10 to see what happens. 💻

**Comment by Neustradamus** [comment]
Good!
What have you seen?

**Comment by sky5walk** [comment]
Attempting to installVersion 3.10.99 (Dec 2 2017) win64 on Windows 10 Pro x64 does not allow for a parallel install of win32?
It prompts user to uninstall the 32-bit version.
I am able to run simultaneous x86/x64 versions of several apps.
Why would DB Browser be different?

**Comment by justinclift** [comment]
@chrisjlocke Ping btw... :smile:

**Comment by sky5walk** [comment]
I use Innosetup and it has '32bit', '64bit' flags. NSIS must have the same. And the destination paths are different on Windows; 'Program Files' vs 'Program Files (x86)'. So, plenty of scriptable checks to add to the installer/uninstaller.

**Comment by chrisjlocke** [comment]
> It prompts user to uninstall the 32-bit version

It should be noted that this is just a prompt.  For most users, this is totally acceptable - they would only want one version.

I run Win 10 x64, and have half a million DB4S installs.  When it prompts, "Do you want to uninstall?", you can select [No] and the installation will continue, asking where you want the install.

If you want x86 and x64 versions of DB4S, then yes, they'll have to be in separate directories - either in c:\program files (and c:\program files (x86) ) or somewhere else.

Obviously there will be an issue when you double click .db / .sql files - which DB4S opens?  You can be clever and create multiple right click options though.

So in conclusion, DB4S **does** allow for a parallel install of win32.
@sky5walk - if you require further details or a screencast, please let me know.

**Comment by justinclift** [comment]
@chrisjlocke Ahhh yeah.  That's a good point.  For the common use case, it does seem to be working how it should.  Well, that saves me some investigating. 😀

**Comment by sky5walk** [comment]
Thanks, I didn't think to click no. I will try that next time.

**Comment by justinclift** [comment]
Looked into the uninstall-not-being-offered problem reported here by @Neustradamus.

### From 3.7.0 (32 or 64 bit)
To 3.7.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.8.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.9.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.1 (32/64 bit) : No prompt to remove the previous version appears
To 3.10.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.10.1 (32/64 bit) : No prompt to remove the previous version appears

### From 3.8.0 (32 or 64 bit)
To 3.7.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.8.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.9.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.1 (32/64 bit) : No prompt to remove the previous version appears
To 3.10.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.10.1 (32/64 bit) : No prompt to remove the previous version appears

### From 3.9.0 (32 or 64 bit)
To 3.7.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.8.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.9.1 (32/64 bit) : No prompt to remove the previous version appears
To 3.10.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.10.1 (32/64 bit) : No prompt to remove the previous version appears

### From 3.9.1 (32 or 64 bit)
To 3.7.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.8.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.1 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.10.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.10.1 (32/64 bit) : A prompt to remove the previous version **does** appear

### From 3.10.0 (32 or 64 bit)
To 3.7.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.8.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.1 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.10.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.10.1 (32/64 bit) : A prompt to remove the previous version **does** appear

### From 3.10.1 (32 or 64 bit)
To 3.7.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.8.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.0 (32/64 bit) : No prompt to remove the previous version appears
To 3.9.1 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.10.0 (32/64 bit) : A prompt to remove the previous version **does** appear
To 3.10.1 (32/64 bit) : A prompt to remove the previous version **does** appear

It looks like the uninstaller option piece is correctly detecting for installed versions 3.9.1 onwards.

My best guess is the uninstaller detection piece is looking for a registry entry matching both the CMake "[CPACK_PACKAGE_VENDOR](https://github.com/sqlitebrowser/sqlitebrowser/blob/4a9f1a2a50f1048f4d1b5d17a43d998082d5935c/CMakeLists.txt#L465)" + "[project](https://github.com/sqlitebrowser/sqlitebrowser/blob/4a9f1a2a50f1048f4d1b5d17a43d998082d5935c/CMakeLists.txt#L1)" variables.

That combination last changed for the 3.9.1 release, which matches up with the above observation:

* https://github.com/sqlitebrowser/sqlitebrowser/commit/aa29f1e4a35e5d253c3b2c705e412bbdec3ea328#diff-af3b638bc2a3e6c650974192a53c7291

They were occasionally changed prior to that too, with the 3.8.0 and 3.9.0 releases.  That matches up with the data above as well, and for why some releases only notice themselves or nearby-previous ones.

For this bug report, I think we're ok from here onwards as long as we don't change the project nor vendor name again.  We'll keep on correctly detecting previous versions, but only for 3.9.1 and onwards.

Figuring out how to add extra detection into the CMake pieces for our older versions... not thinking it'll be worth the effort (with my CMake skills anyway).

Unless someone else thinks it's worthwhile and wants to get it done? :smile:

**Comment by karim** [comment]
Why is this still open? :smiley: Closing it for now.

@Neustradamus Please report back if you still having issues.
